EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs

```diff
@@ -120,33 +117,62 @@ private List<ProteinGroup> ApplyProteinParsimony()
 
             foreach (var kvp in compactPeptideToProteinPeptideMatching)
             {
-                // finds unique peptides (peptides that can belong to only one protein)
-                HashSet<Protein> proteinsAssociatedWithThisPeptide = new HashSet<Protein>(kvp.Value.Select(p => p.Protein));
-                if (proteinsAssociatedWithThisPeptide.Count == 1)
-                {
-                    if (!proteinsWithUniquePeptides.TryGetValue(kvp.Value.First().Protein, out HashSet<PeptideWithSetModifications> peptides))
-                    {
-                        proteinsWithUniquePeptides.Add(kvp.Value.First().Protein, new HashSet<PeptideWithSetModifications>(kvp.Value));
-                    }
-                    else
-                    {
-                        peptides.UnionWith(kvp.Value);
-                    }
-                }
-
                 // if a peptide is associated with a decoy protein, remove all target protein associations with the peptide
                 if (kvp.Value.Any(p => p.Protein.IsDecoy))
-                {
                     kvp.Value.RemoveWhere(p => !p.Protein.IsDecoy);
-                }
 
                 // if a peptide is associated with a contaminant protein, remove all target protein associations with the peptide
                 if (kvp.Value.Any(p => p.Protein.IsContaminant))
-                {
                     kvp.Value.RemoveWhere(p => !p.Protein.IsContaminant);
-                }
             }
 
+            
+            
+            foreach (var kvp in compactPeptideToProteinPeptideMatching)
+            {
+            
+               HashSet<Protein> proteinsAssociatedWithThisPeptide = new HashSet<Protein>(kvp.Value.Select(p => p.Protein));
+               if (proteinsAssociatedWithThisPeptide.Count == 1)
+               {
+                  if (!proteinsWithUniquePeptides.TryGetValue(kvp.Value.First().Protein, out HashSet<PeptideWithSetModifications> peptides))
+                      proteinsWithUniquePeptides.Add(kvp.Value.First().Protein, new HashSet<PeptideWithSetModifications>(kvp.Value));
+                  else
+                      peptides.UnionWith(kvp.Value);
+               }
+               else
+               {
+                    PeptideWithSetModifications[] peptides = new PeptideWithSetModifications[kvp.Value.Count];
+                    int i = 0;
+                    foreach (var peptide in kvp.Value)
+                    {
+                        peptides[i] = peptide;
+                        i++;
+                    }
+
+                    foreach (var peptide in kvp.Value)
+                    {
+                        Protease protease = peptide.digestionParams.Protease;
```

Comment:
// for this peptide with set mods, get its protease
// if there are other peptides with the same protease, it's shared
// if not, it's unique for that protease (and therefore counted as unique for parsimony)